optuna/_callbacks.py

```diff
@@ -55,3 +55,58 @@ def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
         n_complete = len(trials)
         if n_complete >= self._n_trials:
             study.stop()
+
+
+@experimental("2.8.0")
+class RetryFailedTrialCallback:
+    """Retry a failed trial up to a maximum number of times
+
+    When a trail fails, this callback can be used with the :class:`optuna.storage` class to
+    recreate the trial in :obj:`TrialState.WAITING` to queue up the trial to be run again.
+
+    This is helpful in environments where trials may fail due to external conditions, such as
+    being pre-empted by other processes.
+
+    Usage:
+
+        .. testcode::
+
+            from optuna.storages import RetryFailedTrialCallback
+
+            storage = optuna.storages.RDBStorage(
+                args.optuna_storage,
+                heartbeat_interval=60,
+                grace_period=120,
+                failed_trial_callback=RetryFailedTrialCallback(max_retry=3),
+            )
+
+            study = optuna.create_study(
+                storage=storage,
+            )
+
+    """
+
+    def __init__(self, max_retry: int) -> None:
+        self._max_retry = max_retry
+
+    def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
+        user_attrs = {"failed_trial": trial.number}
+        user_attrs.update(trial.user_attrs)
+        failed_trials = (
+            study.user_attrs["failed_trials"] if "failed_trials" in study.user_attrs else []
+        )
+        failed_trials.append(user_attrs["failed_trial"])
+        study.user_attrs.update({"failed_trials": failed_trials})
+        print(study.user_attrs, user_attrs, failed_trials)
+        if failed_trials.count(trial.number) > self._max_retry:
+            return
```

Comment:
Is the user attribute of the study used for counting only? I'm wondering if we cannot get the count without the study user attribute, i.e. by looping over all trials in this method (it's not performance critical so it should be okay). This assumes though that retried trials are labeled in some way.

Reply:
Yes, this is possible, and an improvement. Will change the code accordingly. 🙇

Reply:
Done.